File: Cpp/Game_of_survival/normal_cell.cpp
#include "normal_cell.h"

const std::string& Model::NormalCell::get_alive_state_sign() {
	static std::string obj = "\033[1;31;32m#\033[0m";
	return obj;
}

const std::string& Model::NormalCell::get_dead_state_sign() {
	static std::string obj = " ";
	return obj;
}

Model::NormalCell::NormalCell(bool is_alive) :
	is_alive_now(is_alive),
	CellBase(this->TYPE_CHAR)
{
}

bool Model::NormalCell::is_perceived_as_alive() {
	return this->is_alive_now;
}

void Model::NormalCell::stash_next_state()
{
	int alive_neighbours_count = this->count_alive_neighbours();

	if (this->is_alive_now)
		this->is_alive_next = alive_neighbours_count == 2 || alive_neighbours_count == 3;
	else
		this->is_alive_next = alive_neighbours_count == 3;
}

void Model::NormalCell::release_next_state() {
	this->is_alive_now = this->is_alive_next;
}

const std::string& Model::NormalCell::get_current_state_sign()
{
	return this->is_alive_now ?
		this->get_alive_state_sign() :
		this->get_dead_state_sign();
}

File: Kotlin/src/kotlin/game_of_survival/core/models/cells/NormalCell.kt
package game_of_survival.core.models.cells

import game_of_survival.core.Constants

/**
 * Create a new normal cell, that behaves as in typical Game of Life.
 * @param isAliveNow Is the cell alive at the moment.
 */
class NormalCell(private var isAliveNow: Boolean) : CellBase() {
    private var isAliveNext: Boolean = false
    override val typeChar: Char = Constants.CellTypeChars.NORMAL_CELL
    override fun isPerceivedAsAlive(): Boolean = this.isAliveNow
    override fun stashNextState() {
        val aliveNeighboursCount = this.countAliveNeighbours()
        this.isAliveNext = if (this.isAliveNow) {
            aliveNeighboursCount == 2 || aliveNeighboursCount == 3
        } else {
            aliveNeighboursCount == 3
        }
    }
    override fun releaseNextState() {
        this.isAliveNow = this.isAliveNext
    }
    override fun getCurrentStateSigning(): String =
        if (this.isAliveNow) {
            Constants.CellStateSignings.Normal.ALIVE
        } else {
            Constants.CellStateSignings.Normal.DEAD
        }
}